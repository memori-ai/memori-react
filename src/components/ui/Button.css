.memori-button {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: 4px 15px;
    border: 1px solid var(--memori-button-border-color, #d9d9d9);
    border-radius: var(--memori-button-radius, 5px);
    background: var(--memori-button-bg, #fff);
    box-shadow: var(--memori-button-box-shadow);
    color: var(--memori-button-text, #333);
    font-family: var(--memori-font-family);
    font-size: 14px;
    transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
}
.memori-button--active {
    background: var(--memori-primary);
    color: var(--memori-primary-text);
}

.memori-button:hover {
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.02), 0 4px 8px rgba(0, 0, 0, 0.04);
    color: #333;
    cursor: pointer;
}

.memori-button:hover,
.memori-button:focus {
    border-color: var(--memori-primary);
    color: var(--memori-primary);
}

.memori-button:focus {
    outline-color: var(--memori-primary);
}

.memori-button:active {
    box-shadow: inset var(--memori-button-box-shadow);
}

.memori-button--padded {
    padding: var(--memori-button-padding);
}

.memori-button--block {
    display: block;
    width: 100%;
}

.memori-button--primary {
    background: var(--memori-primary);
    color: var(--memori-primary-text);
}

.memori-button--primary:hover,
.memori-button--primary:focus {
    color: var(--memori-primary-text);
}

.memori-button--primary:active {
    transform: scale(0.95);
}

.memori-button--outlined {
    border-color: var(--memori-primary);
    background: none;
    color: var(--memori-primary);
}

.memori-button--outlined:hover,
.memori-button--outlined:focus {
    border-color: var(--memori-primary);
    background-color: var(--memori-primary);
    color: var(--memori-primary-text);
}

.memori-button--outlined:active {
    transform: scale(0.95);
}

.memori-button--ghost {
    border: 1px solid transparent;
    background: none;
    box-shadow: none;
}

.memori-button--ghost:hover,
.memori-button--ghost:focus {
    border-color: transparent;
    background: rgba(0, 0, 0, 0.04);
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.02), 0 4px 8px rgba(0, 0, 0, 0.04);
    color: inherit;
}

.memori-button--ghost:focus {
    border: 1px solid var(--memori-primary);
}

.memori-button--square {
    border-radius: 0;
}

.memori-button--rounded {
    border-radius: var(--memori-button-radius);
}

.memori-button--danger {
    border-color: var(--memori-error-color);
    background: rgba(0, 0, 0, 0.02);
    color: var(--memori-error-color);
}

.memori-button--danger:hover,
.memori-button--danger:focus {
    border-color: var(--memori-error-color);
    background-color: var(--memori-error-color);
    color: #fff;
}

.memori-button--danger.memori-button--ghost {
    border-color: transparent;
    background: none;
    box-shadow: none;
    color: var(--memori-error-color);
}

.memori-button--danger.memori-button--ghost:hover,
.memori-button--danger.memori-button--ghost:focus {
    border-color: var(--memori-error-color);
    background: rgba(0, 0, 0, 0.02);
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.02), 0 4px 8px rgba(0, 0, 0, 0.04);
    color: var(--memori-error-color);
}

.memori-button--circle {
    padding: 0.75em;
    border-radius: 50%;
    aspect-ratio: 1/1;
}

.memori-button--icon {
    display: inline-flex;
    color: currentColor;
}

.memori-button--icon svg {
    width: 1.25em;
    height: 1.25em;
    fill: currentColor;
    font-size: 1em;
    line-height: 1;
}

.memori-button--icon svg[fill='none'] {
    fill: none;
}

.memori-button--with-icon .memori-button--icon {
    margin-right: 0.5em;
}

.memori-button[disabled] {
    background: var(--memori-button-disabled-bg, #ccc);
    color: var(--memori-button-disabled-color, #000);
    cursor: not-allowed;
    opacity: 0.5;
}
