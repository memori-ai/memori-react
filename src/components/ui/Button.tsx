import React, { FC } from 'react';
import cx from 'classnames';
import Loading from '../icons/Loading';

export interface Props {
  children?:
    | React.ReactChild
    | React.ReactChild[]
    | React.ReactNode
    | JSX.Element;
  primary?: boolean;
  outlined?: boolean;
  ghost?: boolean;
  padded?: boolean;
  block?: boolean;
  icon?: React.ReactNode;
  shape?: 'square' | 'rounded' | 'circle';
  danger?: boolean;
  loading?: boolean;
  disabled?: boolean;
  className?: string;
  title?: string;
  id?: string;
  isActive?: boolean;
  htmlType?: 'button' | 'submit' | 'reset';
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  onMouseDown?: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
  onMouseUp?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  onMouseLeave?: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
  onTouchStart?: (
    event: React.TouchEvent<HTMLButtonElement> | React.MouseEvent
  ) => void;
  onTouchEnd?: (
    event: React.TouchEvent<HTMLButtonElement> | React.MouseEvent
  ) => void;
}

const Button: FC<Props> = ({
  primary = false,
  outlined = false,
  ghost = false,
  padded = true,
  shape = 'rounded',
  danger = false,
  loading = false,
  disabled = false,
  block = false,
  icon,
  className,
  title,
  id,
  htmlType,
  onClick,
  onMouseDown,
  onMouseUp,
  onMouseLeave,
  onTouchStart,
  onTouchEnd,
  children,
  isActive,
}) => (
  <button
    id={id}
    type={htmlType}
    onClick={onClick}
    onMouseDown={onMouseDown}
    onMouseUp={onMouseUp}
    onMouseLeave={onMouseLeave}
    onTouchStart={onTouchStart}
    onTouchEnd={onTouchEnd}
    title={title}
    disabled={loading || disabled}
    className={cx(
      'memori-button',
      {
        'memori-button--primary': primary,
        'memori-button--outlined': outlined,
        'memori-button--ghost': ghost,
        'memori-button--square': shape === 'square',
        'memori-button--rounded': shape === 'rounded',
        'memori-button--circle': shape === 'circle',
        'memori-button--padded': padded,
        'memori-button--block': block,
        'memori-button--with-icon': (icon || loading) && children,
        'memori-button--icon-only': (icon || loading) && !children,
        'memori-button--danger': danger,
        'memori-button--loading': loading,
        'memori-button--active': isActive,
      },
      className
    )}
  >
    {icon && !loading && <span className="memori-button--icon">{icon}</span>}
    {loading && (
      <span className="memori-button--icon loading-icon">
        <Loading loading />
      </span>
    )}
    {children}
  </button>
);

export default Button;
