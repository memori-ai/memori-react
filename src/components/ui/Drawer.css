.memori-drawer {
    position: relative;
    z-index: 10000;
}

.memori-drawer--backdrop {
    position: fixed;
    z-index: 40;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.3);
}

.memori-drawer--container {
    position: fixed;
    z-index: 50;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    overflow-y: auto;
}

.memori-drawer--container-scrollable {
    display: flex;
    min-height: 100%;
    align-items: center;
    justify-content: center;
}

.memori-drawer--panel {
    position: relative;
    width: var(--memori-drawer--width, 100%);
    max-width: 80vw;
    height: 100%;
    min-height: 100vh;
    box-sizing: border-box;
    padding: 1rem 1.5rem 1.5rem;
    margin-right: 0;
    margin-left: auto;
    background-color: #fff;
    font-family: var(--memori-font-family);
}

@media (min-width: 768px) {
    .memori-drawer--panel {
        width: var(--memori-drawer--width--md, 80%);
        max-width: none;
        padding-top: 1.5rem;
        padding-right: 2rem;
        padding-left: 2rem;
    }
}

@media (min-width: 1024px) {
    .memori-drawer--panel {
        width: var(--memori-drawer--width--lg, 60%);
    }
}

.memori-drawer--panel.memori-drawer--with-footer {
    padding-bottom: 0;
}

.memori-drawer--panel .memori-spin {
    display: flex;
    width: 100%;
    height: 100%;
    flex-direction: column;
}

.memori-drawer--panel>.memori-spin {
    margin-top: -1.75rem;
}

.memori-drawer--panel p {
    margin: 0 0 1em;
}

h2.memori-drawer--title {
    margin: 0 0 1rem;
    font-size: 1.5rem;
    font-weight: 700;
}

.memori-drawer--panel-left {
    margin-right: auto;
    margin-left: 0;
}

.memori-drawer--close {
    position: sticky;
    z-index: 70;
    top: 0;
    left: 0;
    padding: 0.5rem;
    margin-top: -1.5rem;
    margin-left: -5.5rem;
}

@media (max-width: 767px) {
    .memori-drawer--close {
        left: 0.5rem;
    }
}

.memori-drawer--close button {
    border-color: #fff;
    color: #fff;
}

@media (max-width: 767px) {
    .memori-drawer--close button {
        color: #000;
    }
}

.memori-drawer--panel-left .memori-drawer--close {
    right: -3.5rem;
    left: auto;
}

@media (max-width: 767px) {
    .memori-drawer--panel-left .memori-drawer--close {
        right: 0.5rem;
        left: auto;
    }
}

.memori-drawer--footer {
    position: sticky;
    bottom: 0;
    left: 0;
    display: flex;
    width: 100%;
    justify-content: flex-end;
    padding-top: 1rem;
    padding-bottom: 1rem;
    border-top: 1px solid rgba(0, 0, 0, 0.25);
    margin-top: auto;
    background-color: #fff;
}

.memori-drawer--footer button+button {
    margin-left: 0.5em;
}

.ease-out {
    transition-timing-function: ease-out;
}

.ease-in {
    transition-timing-function: ease-in;
}

.duration-300 {
    transition-duration: 300ms;
}

.duration-200 {
    transition-duration: 200ms;
}

.opacity-0 {
    opacity: 0;
}

.opacity-100 {
    opacity: 1;
}

.max-w-0 {
    max-width: 0;
}

.max-w-80 {
    max-width: 80%;
}