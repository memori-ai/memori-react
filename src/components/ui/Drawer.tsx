import React, { FC } from 'react';
import { Dialog, Transition } from '@headlessui/react';
import Spin from './Spin';
import Button from './Button';
import Close from '../icons/Close';
import cx from 'classnames';

export interface Props {
  open?: boolean;
  onClose: (value: boolean) => void;
  className?: string;
  title?: string | JSX.Element | React.ReactNode;
  description?: string | JSX.Element | React.ReactNode;
  children: JSX.Element | React.ReactNode;
  footer?: JSX.Element | React.ReactNode;
  loading?: boolean;
  animated?: boolean;
  closable?: boolean;
  side?: 'left' | 'right';
  width?: string;
  widthMd?: string;
  widthLg?: string;
}

const Drawer: FC<Props> = ({
  open = false,
  onClose,
  className,
  title,
  description,
  children,
  footer,
  loading = false,
  animated = true,
  side = 'right',
  closable = true,
  width = '100%',
  widthMd = '80%',
  widthLg = '60%',
}: Props) => (
  <Transition appear show={open} as={React.Fragment}>
    <Dialog
      open={open}
      onClose={onClose}
      className={cx('memori-drawer', className)}
    >
      <Transition.Child
        as={React.Fragment}
        enter="ease-out duration-300"
        enterFrom="opacity-0"
        enterTo="opacity-100"
        leave="ease-in duration-200"
        leaveFrom="opacity-100"
        leaveTo="opacity-0"
      >
        <div className="memori-drawer--backdrop" />
      </Transition.Child>
      <div className="memori-drawer--container">
        <div className="memori-drawer--container-scrollable">
          <Transition.Child
            static
            as={React.Fragment}
            enter="ease-out duration-300"
            enterFrom={animated ? 'max-w-0 opacity-0' : 'opacity-0'}
            enterTo="max-w-80 opacity-100"
            leave="ease-in duration-200"
            leaveFrom="max-w-80 opacity-100"
            leaveTo={animated ? 'max-w-0 opacity-0' : 'opacity-0'}
          >
            <Dialog.Panel
              className={cx('memori-drawer--panel', {
                'memori-drawer--panel-left': side === 'left',
                'memori-drawer--with-footer': !!footer,
              })}
            >
              <style
                dangerouslySetInnerHTML={{
                  __html: `
                    .memori-drawer--panel {
                      --memori-drawer--width: ${width};
                      --memori-drawer--width-lg: ${widthLg};
                      --memori-drawer--width-md: ${widthMd};
                    }
                  `,
                }}
              />
              {closable && (
                <div className="memori-drawer--close">
                  <Button
                    shape="circle"
                    outlined
                    icon={<Close />}
                    onClick={() => onClose(false)}
                  />
                </div>
              )}
              <Spin spinning={loading}>
                {title && (
                  <Dialog.Title className="memori-drawer--title">
                    {title}
                  </Dialog.Title>
                )}
                {description && (
                  <Dialog.Description className="memori-drawer--description">
                    {description}
                  </Dialog.Description>
                )}

                {children}

                {footer && (
                  <div className="memori-drawer--footer">{footer}</div>
                )}
              </Spin>
            </Dialog.Panel>
          </Transition.Child>
        </div>
      </div>
    </Dialog>
  </Transition>
);

export default Drawer;
