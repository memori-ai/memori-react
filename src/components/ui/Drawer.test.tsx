import React from 'react';
import { render } from '@testing-library/react';
import Drawer from './Drawer';
import Button from './Button';

const content = (
  <>
    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>
    <h3>Suspendisse a sodales nulla, sed semper nisi.</h3>
    <p>Proin tincidunt enim in felis aliquet, a ultricies purus bibendum.</p>
    <ul>
      <li>Quisque in ultrices lectus.</li>
      <li>Quisque in ultrices lectus.</li>
      <li>Quisque in ultrices lectus.</li>
    </ul>
    <p>Nulla at urna diam.</p>
  </>
);

const footer = (
  <>
    <Button primary>Ok</Button>
    <Button>Cancel</Button>
  </>
);

beforeEach(() => {
  // @ts-ignore
  window.IntersectionObserver = jest.fn(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
    takeRecords: jest.fn(),
  }));
});

it('renders Drawer unchanged', () => {
  const { container } = render(
    <Drawer open={false} onClose={jest.fn()}>
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer open unchanged', () => {
  const { container } = render(
    <Drawer open={true} onClose={jest.fn()}>
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer with title unchanged', () => {
  const { container } = render(
    <Drawer open={true} onClose={jest.fn()} title="Lorem ipsum">
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer with description unchanged', () => {
  const { container } = render(
    <Drawer
      open={true}
      onClose={jest.fn()}
      description="Lorem ipsum dolor sit amet"
    >
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer loading unchanged', () => {
  const { container } = render(
    <Drawer open={true} onClose={jest.fn()} loading>
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer with footer unchanged', () => {
  const { container } = render(
    <Drawer open={true} onClose={jest.fn()} footer={footer}>
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer non closable unchanged', () => {
  const { container } = render(
    <Drawer open={true} onClose={jest.fn()} closable={false}>
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer side left unchanged', () => {
  const { container } = render(
    <Drawer open={true} onClose={jest.fn()} side="left">
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});

it('renders Drawer with custom widths unchanged', () => {
  const { container } = render(
    <Drawer
      open={true}
      onClose={jest.fn()}
      width="100%"
      widthMd="90%"
      widthLg="80%"
    >
      {content}
    </Drawer>
  );
  expect(container).toMatchSnapshot();
});
