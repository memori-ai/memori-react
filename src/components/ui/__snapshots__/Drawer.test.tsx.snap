// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Drawer loading unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders Drawer non closable unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders Drawer open unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders Drawer side left unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders Drawer unchanged 1`] = `<div />`;

exports[`renders Drawer with custom widths unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders Drawer with description unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders Drawer with footer unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders Drawer with title unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;
